### ISSUE TITLE
Refine the goal of this project

### ISSUE THREAD

**Issue by keaganluttrell** [keaganluttrell/orchestrator#1]
# Goal

To create an orchestration tool for [Unikernels](http://unikernel.org/), like what [Kubernetes](https://kubernetes.io/) is to containers.

## Scope

This will not be a 1:1 feature replication of Kubernetes. It instead will have humble beginnings.  I am still working on the names of things, but for now will keep the conventions very close to Kubernetes

### Nodes
A given application

### Services
A way to group like applications and balance traffic between them

### Gateways
A way to route traffic to given endpoints within the network. Probably `path` based to start

### Manager
A vm with the responsibility of managing Nodes, Services, Gateways and more.

### Needed Features
- CRUD Nodes, Services, Gateways
- Autoscaling for Nodes, maybe services?
- Fault Tolerance -- Be sure failed nodes can be restarted to meet minimums for a given Scaling Policy
- A simple UI to see the given infrastructure
- A Rollback function from the UI
- Helm Like charts to roll out configurations

### Nice To Haves
- Watch Tower type functionality to poll for new VM image releases
- Service Discovery/Dynamic Registry to network Nodes, Services, and Gateways
- Ingress Controller as the only way in to the network.
- Firewall to harden the network.
- TLS Support for all Traffic Serving Nodes (Services, Gateways, etc..)

### Stretch Goals
- Logging and Metrics Built in with UI
- Software Defined Networks or Service Oriented Networking
- An Admin Pane that can get all the metrics, telemetry, health, logs in all one place
- APM for Unikernels
- All of your infrastructure in an easy to see dashboard where you can monitor the health of the entire cluster
- You can navigate to any node, see its performance metrics, health, logs, etc all from within the admin

### Proof of Concept State
- Immutable infrastructure that is defined by configuration.
- Simple by design, just the basics
- Basic Autoscaling
- Basic Routing and Load Balancing
- Basic TLS
- Basic Health Checks
- Basic Node Metrics
- Basic Management of Unikernel VMs
- Basic Admin UI to see into the cluster
- Basic CLI to CRUD Nodes
- Basic Chart of how to build the entire configuration from scratch.
- Someway to periodically save the chart data, perhaps an infrastructure repository in github?

## Roadmap

It all starts with a simulation. In order to keep complexity to a minimum, this application will start as a simulation of Nodes. The goal is to prove the concept first and iron out basic provisioning headaches with in memory objects that will represent Nodes in the full feature repo. Once the proof of concept is complete, I will try to create a minimum viable product with unikernels deploying a very basic application that serves a website.  My hope from here is to start to add some services to the site that display the functionality of load balancing and gateway routing.
